File: py/candlestick_patterns/patterns/harami_cross.py
"""Harami Cross pattern (2-candle reversal)."""
from __future__ import annotations

from ..core.primitives import real_body
from ...fuzzy import t_product_all


def harami_cross(self) -> float:
    """Harami Cross: a two-candle reversal pattern.

    Like Harami, but the second candle is a doji instead of just short.

    Must have:
    - first candle: long real body,
    - second candle: doji body contained within the first candle's real body.

    Category B: direction from 1st candle color (continuous).

    Returns:
        Continuous float in [-100, +100].
    """
    if not self._enough(2, self._long_body, self._doji_body):
        return 0.0

    o1, h1, l1, c1 = self._bar(2)
    o2, h2, l2, c2 = self._bar(1)

    # Fuzzy size conditions.
    mu_long1 = self._mu_greater(real_body(o1, c1), self._long_body, 2)
    mu_doji2 = self._mu_less(real_body(o2, c2), self._doji_body, 1)

    # Fuzzy containment: 1st body encloses 2nd body.
    eq_avg = self._avg(self._equal, 1)
    eq_width = self._fuzz_ratio * eq_avg if eq_avg > 0.0 else 0.0

    mu_enc_upper = self._mu_ge_raw(max(o1, c1), max(o2, c2), eq_width)
    mu_enc_lower = self._mu_lt_raw(min(o1, c1), min(o2, c2), eq_width)

    confidence = t_product_all(mu_long1, mu_doji2, mu_enc_upper, mu_enc_lower)

    # Direction: opposite of 1st candle color.
    direction = -1.0 if c1 >= o1 else 1.0

    return direction * confidence * 100.0

File: ts/candlestick-patterns/patterns/harami-cross.ts
/** Harami Cross pattern (2-candle reversal). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll } from '../../fuzzy/index.ts';
import { realBody } from '../core/primitives.ts';

/**
 * Harami Cross: a two-candle reversal pattern.
 *
 * Like Harami, but the second candle is a doji instead of just short.
 *
 * Must have:
 * - first candle: long real body,
 * - second candle: doji body contained within the first candle's real body.
 *
 * Category B: direction from 1st candle color (continuous).
 *
 * Returns:
 *     Continuous number in [-100, +100].
 */
export function haramiCross(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(2, cp.longBody, cp.dojiBody)) return 0.0;

    const b1 = cp.bar(2);
    const b2 = cp.bar(1);

    // Fuzzy size conditions.
    const muLong1 = cp.muGreaterCS(realBody(b1.o, b1.c), cp.longBody, 2);
    const muDoji2 = cp.muLessCS(realBody(b2.o, b2.c), cp.dojiBody, 1);

    // Fuzzy containment: 1st body encloses 2nd body.
    const eqAvg = cp.avgCS(cp.equal, 1);
    let eqWidth = cp.fuzzRatio * eqAvg;
    if (eqAvg <= 0.0) eqWidth = 0.0;

    const muEncUpper = cp.muGeRaw(Math.max(b1.o, b1.c), Math.max(b2.o, b2.c), eqWidth);
    const muEncLower = cp.muLtRaw(Math.min(b1.o, b1.c), Math.min(b2.o, b2.c), eqWidth);

    const confidence = tProductAll(muLong1, muDoji2, muEncUpper, muEncLower);
    // Direction: opposite of 1st candle color.
    const direction = b1.c < b1.o ? 1.0 : -1.0;
    return direction * confidence * 100.0;
}
